packages/server-wallet/src/protocols/channel-closer.ts

```diff
@@ -39,142 +38,137 @@ export class ChannelCloser {
   public async crank(objective: DBCloseChannelObjective, response: WalletResponse): Promise<void> {
     const channelToLock = objective.data.targetChannelId;
 
-    let attemptAnotherProtocolStep = true;
+    await this.store.lockApp(channelToLock, async tx => {
+      const channel = await this.store.getChannel(channelToLock, tx);
+      if (!channel) {
+        throw new Error('Channel must exist');
+      }
+      try {
+        if (!ensureAllAllocationItemsAreExternalDestinations(channel)) {
+          return;
+        }
 
-    while (attemptAnotherProtocolStep) {
-      await this.store.lockApp(channelToLock, async tx => {
-        const protocolState = await this.getCloseChannelProtocolState(
-          this.store,
-          objective.data.targetChannelId,
-          tx
-        );
-        try {
-          let turnNum;
-          if (!ensureAllAllocationItemsAreExternalDestinations(protocolState)) {
-            attemptAnotherProtocolStep = false;
-          } else if ((turnNum = turnNumberToSign(protocolState))) {
-            await this.signState(protocolState, turnNum, tx, response);
-          } else if (shouldDefund(protocolState)) {
-            await this.defund(protocolState, tx);
-          } else if (shouldCompleteObjective(protocolState)) {
-            attemptAnotherProtocolStep = false;
-            await this.completeObjective(objective, protocolState, tx, response);
-          } else {
-            response.queueChannelState(protocolState.app);
-            attemptAnotherProtocolStep = false;
-          }
-        } catch (error) {
-          this.logger.error({error}, 'Error taking a protocol step');
-          await tx.rollback(error);
-          attemptAnotherProtocolStep = false;
+        const turnNum = turnNumberToSign(channel);
+        // Explicitely check undefined just in case we want to sign turnNum 0
+        if (turnNum !== undefined) {
+          await this.signState(channel, turnNum, tx, response);
```

Comment:
Should there be a check to ensure that we haven't already done this on a previous crank?

Reply:
@tomclose I think you are pointing out that the implementation of the `crank` function could be clearer. I think the approach in f69f89c should help.